libraries/AP_Proximity/AP_Proximity_Boundary_3D.h

```diff
@@ -0,0 +1,147 @@
+	    bool operator ==(const Face &other) const { return ((layer == other.layer) && (sector == other.sector)); }
+	    bool operator !=(const Face &other) const { return ((layer != other.layer) || (sector != other.sector)); }
+
+        uint8_t layer;  // vertical "steps" on the 3D Boundary. 0th layer is the bottom most layer, 1st layer is 30 degrees above (in body frame) and so on
+        uint8_t sector; // horizontal "steps" on the 3D Boundary. 0th sector is directly in front of the vehicle. Each sector is 45 degrees wide.
+    };
+
+    // returns face corresponding to the provided yaw and (optionally) pitch
+    // pitch is the vertical body-frame angle (in degrees) to the obstacle (0=directly ahead, 90 is above the vehicle?)
+    // yaw is the horizontal body-frame angle (in degrees) to the obstacle (0=directly ahead of the vehicle, 90 is to the right of the vehicle)
+    Face get_face(float pitch, float yaw) const;
+    Face get_face(float yaw) const { return get_face(0, yaw); }
+
+    // Set the actual body-frame angle(yaw), pitch, and distance of the detected object.
+    // This method will also mark the sector and layer to be "valid",
+    // This distance can then be used for Obstacle Avoidance
+    // Assume detected obstacle is horizontal (zero pitch), if no pitch is passed
+    void set_face_attributes(const Face &face, float pitch, float yaw, float distance);
+    void set_face_attributes(const Face &face, float yaw, float distance) { set_face_attributes(face, 0, yaw, distance); }
+
+    // add a distance to the boundary if it is shorter than any other provided distance since the last time the boundary was reset
+    // pitch and yaw are in degrees, distance is in meters
+    void add_distance(float pitch, float yaw, float distance);
+    void add_distance(float yaw, float distance) { add_distance(0, yaw, distance); }
+
+    // update boundary points used for simple avoidance based on a single sector and pitch distance changing
+    //   the boundary points lie on the line between sectors meaning two boundary points may be updated based on a single sector's distance changing
+    //   the boundary point is set to the shortest distance found in the two adjacent sectors, this is a conservative boundary around the vehicle
+    void update_boundary(const Face &face);
+
+    // update middle layer boundary points
+    void update_middle_boundary();
+
+    // reset boundary.  marks all distances as invalid
+    void reset();
+
+    // Reset this location, specified by Face object, back to default
+    // i.e Distance is marked as not-valid
+    void reset_face(const Face &face);
+
+    // get distance for a face.  returns true on success and fills in distance argument with distance in meters
+    bool get_distance(const Face &face, float &distance) const;
+
+    // Get the total number of obstacles 
+    uint8_t get_obstacle_count() const;
+
+    // Appropriate layer and sector are found from the passed obstacle_num
```

Comment:
It would be nice if the first line of this comment gave the high level view of what the method does.  The comments with more details of how it's done could be moved to the .cpp file.

I'd shorten the comment to something more like, "return a body frame vector (in cm) to an obstacle.  FLT_MAX is returned if the obstacle_num provided does not produce a valid obstacle".